### PR TITLE
feat(file source): Support compressed files

### LINKED ISSUE #848: Add support for gzipped files to file source
We have a few pipelines which read gzipped files, transform them and send the result to Elasticsearch. It would be great to have the same functionality in vector.

In Logstash, this is done automatically if the input file is of type `gzip`.
For further reference, [here](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-file.html#plugins-inputs-file-mode) is the logstash file input plugin that supports gzip.

### PR DESCRIPTION
**Pull request by mre**
This adds support for gzip and zip files in the file source.
Closes #848.

### CODE REVIEW

**Comment by mre** [comment]
We should add some tests before merging this.
I'm thinking about adding an integration test to the `tests` folder for this use-case if no one objects .

**Comment by mre** [comment]
Coming back to this, I noticed that compression is not working yet.
I get the compressed output stream instead of the uncompressed one.
Might be related to `read_until_with_max_size`, but I doubt it.
Maybe it's just something obvious that I can't spot at the moment...

File: lib/file-source/src/file_watcher.rs [hunk 1]
```diff
@@ -51,23 +53,35 @@ impl FileWatcher {
...
             inode: metadata.ino(),
             is_dead: false,
         })
     }
 
+    // Depending on the file type we create a different type of reader.
+    // This allows us to handle compressed files as well as uncompressed files.
+    fn wrap_reader_for_filetype(reader: io::BufReader<File>, path: &PathBuf) -> Box<dyn BufRead> {
+        match path.extension().and_then(OsStr::to_str) {
+            Some("gzip") => Box::new(io::BufReader::new(MultiGzDecoder::new(reader))),
```
Comment (lukesteensen) [inline lib/file-source/src/file_watcher.rs]: Probably want this to be `Some("gzip") | Some("gz") => ...`

Comment (mre) [reply lib/file-source/src/file_watcher.rs]: True. Changed. 👍

**Comment by binarylogic** [comment]
@mre do you need another review for this PR? Let us know when you're done, we'll review and merge. Thanks!

**Comment by mre** [comment]
The main problem is, that decompression does not work as expected. I still get compressed output when I try to print the messages to stdout. 😕
Not sure what's going on there, so I'm thankful for any hint.

**Comment by binarylogic** [comment]
@mre thanks for working on this. Any update on the progress? Do you want to make the suggested changes? Otherwise we can have someone internal take this to the finish line.

**Comment by binarylogic** [comment]
Thanks @mre. We'll have @lukesteensen take this to the finish line in the next couple of weeks.

**Pull request closed without merging**